packages/hash/frontend/src/components/BlockLoader/data-map-editor.tsx

```diff
@@ -0,0 +1,357 @@
+          sx={({ palette }) => ({
+            display: "flex",
+            borderBottom: `1px solid ${palette.gray[30]}`,
+            py: 1,
+          })}
+        >
+          <Box width="30%">
+            <Typography fontWeight={600} variant="smallTextLabels">
+              Target path
+            </Typography>
+          </Box>
+          <Box width="30%">
+            <Typography fontWeight={600} variant="smallTextLabels">
+              Source path
+            </Typography>
+          </Box>
+          <Box width="30%" ml={4}>
+            <Typography fontWeight={600} variant="smallTextLabels">
+              Default
+            </Typography>
+          </Box>
+        </Box>
+
+        {targetPaths.map((targetPath) => {
+          const { required, schema: targetPathSchema } = schemaByPathParts(
+            targetSchema,
+            targetPath.split("."),
+          );
+
+          const { description, type } = targetPathSchema;
+
+          return (
+            <Box
+              key={targetPath}
+              sx={({ palette }) => ({
+                py: 1,
+                display: "flex",
+                alignItems: "center",
+                borderBottom: `1px solid ${palette.gray[20]}`,
+              })}
+            >
+              <Box width="30%">
+                <Typography fontWeight={500} variant="smallTextLabels">
+                  {targetPath}
+                  {required ? "*" : ""}
+                  <Typography fontWeight={300} ml={1} variant="microText">
+                    {Array.isArray(type) ? type.join(", ") : type}
+                  </Typography>
+                </Typography>
+                <Typography
+                  sx={({ palette }) => ({
+                    color: palette.gray[50],
+                  })}
+                  fontWeight={300}
+                  component="p"
+                  variant="microText"
+                >
+                  {description ?? "No description provided"}
+                </Typography>
+              </Box>
+              <Box width="30%">
+                <Select
+                  size="small"
+                  defaultValue={
+                    schemaMap?.transformations?.[targetPath]?.sourceKey ??
+                    "__nothing"
+                  }
+                  onChange={(evt) =>
+                    updateSchemaMap((existingMap) => ({
+                      ...(existingMap ?? { mapId }),
+                      transformations: {
+                        ...existingMap.transformations,
+                        [targetPath]: {
+                          default:
+                            existingMap.transformations?.[targetPath]?.default,
+                          sourceKey:
+                            evt.target.value === "__nothing"
+                              ? undefined
+                              : evt.target.value,
+                        },
+                      },
+                    }))
+                  }
```

Comment:
Seems like something that could be made significantly more readable with `Immer`?

Reply:
https://github.com/hashintel/hash/pull/918/commits/5080167c62a147a2dc87c0c9bdb0bc9be505001b